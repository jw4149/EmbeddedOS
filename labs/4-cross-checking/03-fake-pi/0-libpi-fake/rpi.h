#ifndef __RPI_H__
#define __RPI_H__
// simplest way to rewrap --- just define replacements in header --- don't
// even need .c
//
// easy, but two downsides: 
//   1. we are polluting the pi program namespace with system includes.
//   2. we aren't using the raw rpi.h so can get different behavior.
#include <stdio.h>
#include <stdlib.h>

// redefine printk as fprintf using variadic macros.
#define printk(args...) fprintf(stderr, args)

static void inline clean_reboot(void) {
    printk("DONE!!!\n");
    exit(0);
}

void notmain(void);
#endif
