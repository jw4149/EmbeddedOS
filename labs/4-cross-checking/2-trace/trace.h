#ifndef __TRACE_H__
#define __TRACE_H__
// engler, cs140e: dirt simple tracing system.

// start tracing : 
//   <buffer_p> = 0: tells the code to immediately print the 
//                   trace.  
//   <buffer_p> != 0 defers the printing until <trace_stop> is called.
//
//  is an error: if you were already tracing.
void trace_start(int buffer_p);

// stop tracing
//  - error: if you were not already tracing.
void trace_stop(void);


// pause tracing: pause tracing, resume after calling
// trace_resume();
//
// only useful if you were running with buffering.
void trace_pause(void);
void trace_resume(void);


// make the above nicer so we can easily trace a function/chunk of code.
#define stringify(_x) #_x

#define trace_fn(fn) do {                                    \
    printk("about to trace: <%s>\n", stringify(fn));        \
    trace_start(0);                                         \
    fn;                                                     \
    trace_stop();                                           \
    printk("done: <%s>\n", stringify(fn));                  \
} while(0)

#endif
