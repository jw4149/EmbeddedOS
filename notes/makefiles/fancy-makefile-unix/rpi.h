#include <stdio.h>

#define printk printf
#define debug printf
