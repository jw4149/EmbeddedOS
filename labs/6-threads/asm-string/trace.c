// very cool.
#include "rpi.h"

// have backtrace.

void __cyg_profile_func_enter (void *this_fn, void *call_site) 
  __attribute__((no_instrument_function));
void __cyg_profile_func_exit  (void *this_fn, void *call_site) 
  __attribute__((no_instrument_function));

void __cyg_profile_func_enter (void *this_fn, void *call_site) 
{
    static int in_enter;

    if(in_enter)
        return;

    in_enter = 1;
    printk("tracing %x\n", this_fn);
    in_enter = 0;
}

void __cyg_profile_func_exit  (void *this_fn, void *call_site) 
{
    return;
}
