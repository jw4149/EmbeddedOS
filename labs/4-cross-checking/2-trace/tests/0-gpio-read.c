// example client that uses our tracing system: can cross check these values
// against the ones on your laptop!
#include "rpi.h"
#include "trace.h"

void notmain(void) {
    uart_init();

    printk("about to trace gpio_read(20):\n");
    trace_start(0);
    int v = gpio_read(20);
    trace_stop();
    printk("done: gpio_read(20)=%d\n", v);

    clean_reboot();
}
