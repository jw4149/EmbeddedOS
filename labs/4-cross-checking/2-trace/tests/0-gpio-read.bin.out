TRACE:GET32(0x0x20200034)=0x0xf00081ff
